### ISSUE TITLE
Support gradient clipping by norm with FSDP

### ISSUE THREAD

**Issue by awaelchli** [Lightning-AI/pytorch-lightning#19235]
### Description & Motivation

Our current implementation of gradient clipping for FSDP is limited to clipping by value only. Norm is not supported:
https://github.com/Lightning-AI/pytorch-lightning/blob/f75f3bc1c65a2bb3388ffd68b1ee70ed9741525c/src/lightning/pytorch/plugins/precision/fsdp.py#L77-L84

The reason is that clipping by norm needs to be called through the FSDP API and this wasn't realized in Lightning yet, because it can't be done directly through the optimizer (the FSDP module reference is required): https://pytorch.org/docs/stable/fsdp.html#torch.distributed.fsdp.FullyShardedDataParallel.clip_grad_norm_

### Pitch

Support clipping by norm.

Change the API from 

```
class Precision:
    ...
    def clip_grad_by_norm(self, optimizer: Optimizer, clip_val: Union[int, float]) -> None:
        ...
```

to 

```
class Precision:
    ...
    def clip_grad_by_norm(self, module: Module, optimizer: Optimizer, clip_val: Union[int, float]) -> None:
        ...
```
to take the module as input. The implementation in `FSDPPrecision` would then call `module.clip_grad_norm()` instead of `torch.nn.utils.clip_grad_norm_`. 

The `LightningModule.clip_gradients()` method should then pass `self.trainer.model` to `self.trainer.precision_plugin.clip_gradients()`. 

### Alternatives

There is not much else we can do. I believe the proposal above will lead to the least amount of breaking changes (only affects the signature of the precision plugin methods.

**Comment by xin-w8023** [comment]
Any updates about this?

**Comment by amorehead** [comment]
Agreed, any updates?

**Comment by amorehead** [comment]
Based on my experience, clipping by gradient norms can be a crucial component of successfully training a large (model-parallel) network. It seems Hugging Face's `accelerate` library already supports this [functionality](https://huggingface.co/docs/accelerate/main/en/package_reference/accelerator#accelerate.Accelerator.clip_grad_norm_) for FSDP, and PyTorch's `FullyShardedDataParallel` object even has a `clip_grad_norm_` method attached to it to facilitate rank-aware gradient norm-clipping. Are there any particular gotchas that one needs to keep in mind when wiring this functionality into the `LightningModule` class?

**Comment by amorehead** [comment]
I've taken a shot at implementing a PR for this issue [here](https://github.com/Lightning-AI/pytorch-lightning/pull/20784). Open to any and all feedback on it.
